network/native_endpointclient_linux.go

```diff
@@ -0,0 +1,438 @@
+}
+
+type NativeEndpointClient struct {
+	eth0VethName      string //So like eth0
+	ethXVethName      string //So like eth0.X
+	vnetVethName      string //Peer is containerVethName
+	containerVethName string //Peer is vnetVethName
+
+	vnetMac      net.HardwareAddr
+	containerMac net.HardwareAddr
+	ethXMac      net.HardwareAddr
+
+	vnetNSName           string
+	vnetNSFileDescriptor uintptr
+
+	mode           string
+	vlanID         int
+	netnsClient    netns.NetnsInterface
+	netlink        netlink.NetlinkInterface
+	netioshim      netio.NetIOInterface
+	plClient       platform.ExecClient
+	netUtilsClient networkutils.NetworkUtils
+}
+
+func NewNativeEndpointClient(
+	eth0VethName string,
+	ethXVethName string,
+	vnetVethName string,
+	containerVethName string,
+	vnetNSName string,
+	mode string,
+	vlanid int,
+	nl netlink.NetlinkInterface,
+	plc platform.ExecClient,
+) *NativeEndpointClient {
+	log.Printf("[native] Create new native client: eth0:%s, ethX:%s, vnet:%s, cont:%s, id:%s",
+		eth0VethName, ethXVethName, vnetVethName, containerVethName, vlanid)
+	client := &NativeEndpointClient{
+		eth0VethName:      eth0VethName,
+		ethXVethName:      ethXVethName,
+		vnetVethName:      vnetVethName,
+		containerVethName: containerVethName,
+		vnetNSName:        vnetNSName,
+		mode:              mode,
+		vlanID:            vlanid,
+		netnsClient:       netns.NewNetns(),
+		netlink:           nl,
+		netioshim:         &netio.NetIO{},
+		plClient:          plc,
+		netUtilsClient:    networkutils.NewNetworkUtils(nl, plc),
+	}
+
+	return client
+}
+
+// Adds interfaces to the vnet (created if not existing) and vm namespace
+func (client *NativeEndpointClient) AddEndpoints(epInfo *EndpointInfo) error {
+	var err error
+	// VM Namespace
+	err = client.PopulateVM(epInfo)
+	if err != nil {
+		return err
+	}
+	// VNET Namespace
+	err = ExecuteInNS(client.vnetNSName, epInfo, client.PopulateVnet)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Called from AddEndpoints, Namespace: VM
+func (client *NativeEndpointClient) PopulateVM(epInfo *EndpointInfo) error {
+	vmNS, err := client.netnsClient.Get()
+	if err != nil {
+		return newErrorNativeEndpointClient("Failed to get VM NS handle", err.Error())
+	}
+	// Inform current namespace
+	returnedTo, err := GetCurrentThreadNamespace()
+	if err != nil {
+		return newErrorNativeEndpointClient("Failed to get VM NS", err.Error())
+	} else {
+		log.Printf("[native] VM Namespace: %s", returnedTo.file.Name())
+	}
+
+	log.Printf("[native] Checking if NS exists...")
+	var existingErr error
+	vnetNS, existingErr := client.netnsClient.GetFromName(client.vnetNSName)
+	// If the ns does not exist, the below code will trigger to create it
+	if existingErr != nil {
+		if !strings.Contains(strings.ToLower(existingErr.Error()), "no such file or directory") {
+			return newErrorNativeEndpointClient("Error other than vnet NS doesn't exist", existingErr.Error())
```

Comment:
- errors should not be `Sentence case`, they tend to be embedded as logs and should be readable as fragments
- should use error wrapping